### PR TITLE
Interaction - Add "pass grenade" interaction

### PR DESCRIPTION
**Pull request by Timi007**
**When merged this pull request will:**
- Add "pass grenade" interaction that passes the currently selected grenade

screenshot2

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: addons/interaction/functions/fnc_getThrowableInfo.sqf
Comment (johnb432) [inline addons/interaction/functions/fnc_getThrowableInfo.sqf:1]: Do we really need this? What's the benefit?

**Review by johnb432 — COMMENTED** [review]
> When selecting the category, it will either use the last grenade that was passed or the next available one in the inventory.

In my personal opinion, that particular aspect of the PR is not intuitive (magazine passing doesn't have that, so why should grenade passing have it? also no documentation), not helpful (it only saves you the final mouse movement, so it save you pretty little; it's unlikely you'd be passing numerous grenades this way, you'd probably put it in a backpack or the ground) and therefore not useful.

I'd like to hear more from other members of the ACE team though, before you undertake any changes related to that.

File: addons/interaction/functions/fnc_getThrowableInfo.sqf
Comment (Timi007) [reply addons/interaction/functions/fnc_getThrowableInfo.sqf:1]: Do you mean the whole function? It is a helper function to reduce duplicate code and cache config properties.

**Comment by Timi007** [comment]
> In my personal opinion, that particular aspect of the PR is not intuitive (magazine passing doesn't have that, so why should grenade passing have it? also no documentation), not helpful (it only saves you the final mouse movement, so it save you pretty little; it's unlikely you'd be passing numerous grenades this way, you'd probably put it in a backpack or the ground) and therefore not useful.

Do I understand you correctly that you don't want to have this split into lethal and non-lethal grenades? Should all grenades be listed under "Pass grenade" directly?

**Comment by Drofseh** [comment]
> > When selecting the category, it will either use the last grenade that was passed or the next available one in the inventory.
> 
> In my personal opinion, that particular aspect of the PR is not intuitive (magazine passing doesn't have that, so why should grenade passing have it? also no documentation), not helpful (it only saves you the final mouse movement, so it save you pretty little; it's unlikely you'd be passing numerous grenades this way, you'd probably put it in a backpack or the ground) and therefore not useful.
> 
> I'd like to hear more from other members of the ACE team though, before you undertake any changes related to that.

I kinda feel like it's too busy with a node for every grenade type carried and that a single pass grenade node that just passes your currently selected grenade would be fine.

File: addons/interaction/functions/fnc_getThrowableInfo.sqf
Comment (johnb432) [reply addons/interaction/functions/fnc_getThrowableInfo.sqf:1]: I meant the cache itself - is there a noticeable difference between caching and not?

**Comment by johnb432** [comment]
> When passing a magazine, you don't select a specific type

Yea, you're right, forgot about that despite having the code under my eyes.

> If needed, I can document this feature in the wiki.

Regardless of what will be done in the end, some documentation should be provided.

> I kinda feel like it's too busy with a node for every grenade type carried and that a single pass grenade node that just passes your currently selected grenade would be fine.

I agree with it being too busy, especially if you use the radial menu, but I do see the appeal of being able to pass any grenade.

**Comment by Timi007** [comment]
So do all agree to have only one "pass current grenade" action node and to pass the currently selected grenade?

**Comment by Timi007** [comment]
Ready for review

File: addons/interaction/functions/fnc_passThrowable.sqf [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+TRACE_4("Pass throwable params",_player,_target,_throwable,_animate);
+
+if (isNull _target) exitWith {ERROR("Target is null.")};
+if (isNull _player) exitWith {ERROR("Player is null.")};
+if (_throwable isEqualTo "") exitWith {ERROR("No throwable specified.")};
+if !([_target, _throwable] call CBA_fnc_canAddItem) exitWith {ERROR("Cannot add throwable to target due to lack of inventory space.")};
+
+if (_animate) then {[_player, "PutDown"] call EFUNC(common,doGesture)};
+
+_player removeItem _throwable;
```
Comment (johnb432) [inline addons/interaction/functions/fnc_passThrowable.sqf:31]: The condition is evaluated only once a second, which allows up to 1 second where you could have no grenades but still be able to pass a grenade.
The fix is to check if the item was successfully removed and stop before doing anything else:

```suggestion
if !([_player, _throwable] call CBA_fnc_removeItem) exitWith {};

if (_animate) then {[_player, "PutDown"] call EFUNC(common,doGesture)};
```

File: addons/interaction/functions/fnc_canPassThrowable.sqf [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Checks if target unit can accept given throwable.
+ * Does not check if the throwable exists in the inventory of the player.
+ *
+ * Arguments:
+ * 0: Unit to pass the throwable to <OBJECT>
+ * 1: Unit that passes the throwable <OBJECT>
+ * 2: Throwable classname <STRING>
+ *
+ * Return Value:
+ * Unit can pass throwables <BOOL>
```
Comment (johnb432) [inline addons/interaction/functions/fnc_canPassThrowable.sqf:13]:
```suggestion
 * Unit can pass throwable <BOOL>
```

File: addons/interaction/functions/fnc_passThrowable.sqf [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+_target addItem _throwable;
```
Comment (johnb432) [inline addons/interaction/functions/fnc_passThrowable.sqf:32]: Grenades can have more than 1 ammo (see SOG's `vn_v40_grenade_mag`) - you need to take that into account.

**Review by johnb432 — CHANGES_REQUESTED** [review]

File: addons/interaction/functions/fnc_passThrowable.sqf (hunk 3 above)
Comment (Timi007) [reply addons/interaction/functions/fnc_passThrowable.sqf]: Done

File: addons/interaction/functions/fnc_passThrowable.sqf [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [_target, _player, "HandGrenade"] call ace_interaction_fnc_passThrowable
+ *
+ * Public: No
+ */
+
+params [["_target", objNull, [objNull]], ["_player", objNull, [objNull]], ["_throwable", "", [""]], ["_animate", true, [true]]];
```
Comment (johnb432) [inline addons/interaction/functions/fnc_passThrowable.sqf:21]: No need for default arguments, as this isn't API.
```suggestion
params ["_player", "_target", "_throwable", ["_animate", true, [true]]];
```

File: addons/interaction/functions/fnc_canPassThrowable.sqf [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Timi007
+ * Checks if target unit can accept given throwable.
+ * Does not check if the throwable exists in the inventory of the player.
+ *
+ * Arguments:
+ * 0: Unit to pass the throwable to <OBJECT>
+ * 1: Unit that passes the throwable <OBJECT>
```
Comment (johnb432) [inline addons/interaction/functions/fnc_canPassThrowable.sqf:9]:
```suggestion
 * 0: Unit that passes the throwable <OBJECT>
 * 1: Unit to pass the throwable to <OBJECT>
```

File: addons/interaction/functions/fnc_passThrowable.sqf [hunk 6]
```diff
@@ -0,0 +1,59 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Timi007
+ * Pass throwable to another unit.
+ *
+ * Arguments:
+ * 0: Unit to pass the throwable to <OBJECT>
+ * 1: Unit that passes the throwable <OBJECT>
```
Comment (johnb432) [inline addons/interaction/functions/fnc_passThrowable.sqf:8]:
```suggestion
 * 0: Unit that passes the throwable <OBJECT>
 * 1: Unit to pass the throwable to <OBJECT>
```

File: addons/interaction/functions/fnc_passThrowable.sqf [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ * 2: Throwable classname <STRING>
+ * 3: Play passing animation <BOOL> (default: true)
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [_target, _player, "HandGrenade"] call ace_interaction_fnc_passThrowable
```
Comment (johnb432) [inline addons/interaction/functions/fnc_passThrowable.sqf:16]:
```suggestion
 * [_player, _target, "HandGrenade"] call ace_interaction_fnc_passThrowable
```

File: addons/interaction/functions/fnc_canPassThrowable.sqf [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * 2: Throwable classname <STRING>
+ *
+ * Return Value:
+ * Unit can pass throwable <BOOL>
+ *
+ * Example:
+ * [_target, _player, "HandGrenade"] call ace_interaction_fnc_canPassThrowable
```
Comment (johnb432) [inline addons/interaction/functions/fnc_canPassThrowable.sqf:16]:
```suggestion
 * [_player, _target, "HandGrenade"] call ace_interaction_fnc_canPassThrowable
```

**Review by johnb432 — APPROVED** [review]

**Review by PabstMirror — APPROVED** [review]

**Pull request merged**